src/server/pkg/hashtree/db.go

```diff
@@ -556,45 +559,57 @@ func (h *dbHashTree) Deserialize(_r io.Reader) error {
 					}
 				}
 				return nil
-			}); err != nil {
-				return err
+			}); err != nil || copyCtx.Err() != nil {
+				return err // may return nil if copyCtx was closed
 			}
 			if count <= 0 {
 				return nil
 			}
 		}
 	})
-	for {
-		hdr.Reset()
-		if err := r.Read(hdr); err != nil {
-			if err == io.EOF {
-				break
-			}
-			return err
-		}
-		bucket := b(hdr.Bucket)
-		kvs <- &keyValue{nil, bucket}
+	eg.Go(func() error {
+		defer close(kvs)
 		for {
-			_k, err := r.ReadBytes()
-			if err != nil {
+			hdr.Reset()
+			// TODO(msteffen): don't block on Read if copyCtx() is cancelled?
+			if err := r.Read(hdr); err != nil {
+				if err == io.EOF {
+					break
+				}
 				return err
 			}
-			if bytes.Equal(_k, SentinelByte) {
-				break
+			bucket := b(hdr.Bucket)
+			select {
+			case kvs <- &keyValue{nil, bucket}:
+			case <-copyCtx.Done():
+				return nil
 			}
-			// we need to make copies of k and v because the memory will be reused
-			k := make([]byte, len(_k))
-			copy(k, _k)
-			_v, err := r.ReadBytes()
-			if err != nil {
-				return err
+			for {
+				_k, err := r.ReadBytes()
+				if err != nil {
+					return err
+				}
+				if bytes.Equal(_k, SentinelByte) {
+					break
+				}
+				// we need to make copies of k and v because the memory will be reused
+				k := make([]byte, len(_k))
+				copy(k, _k)
+				_v, err := r.ReadBytes()
+				if err != nil {
+					return err
+				}
+				v := make([]byte, len(_v))
+				copy(v, _v)
+				select {
+				case kvs <- &keyValue{k, v}:
+				case <-copyCtx.Done():
+					return nil
+				}
 			}
-			v := make([]byte, len(_v))
-			copy(v, _v)
-			kvs <- &keyValue{k, v}
 		}
-	}
-	close(kvs)
+		return nil
+	})
```

Comment:
@brycemcanally do these changes look correct to you? They look reasonable to me but you know the code better.

Reply:
Yeah, they seem fine.